src/trace_queue.js

```diff
@@ -0,0 +1,188 @@
+    constructor(batchSender) {
+        super();
+        this.batchSize = config.getConfig().batchSize;
+        this.batchBytesLimit = consts.batchBytesLimit;
+        this.traces = [];
+        this.batchQueueTime = config.getConfig().batchQueueTime;
+        this.batchSender = batchSender;
+        this.currentByteSize = 0;
+        this.initQueue();
+    }
+
+    /**
+   * Init queue event listners
+   */
+    initQueue() {
+        this.removeAllListeners();
+        this.currentByteSize = 0;
+        this.flush();
+        //  check if queue reached batch size
+        this.on('traceQueued', () => {
+            if (this.batchSizeReached()) {
+                utils.debugLog(`Queue size reached ${this.currentSize}, batch ready! Releasing... `);
+                this.releaseBatch();
+            }
+        });
+
+        // update queue size in bytes + check if limit reached
+        this.on('traceQueued', (trace) => {
+            this.addToCurrentByteSize(trace);
+            if (this.byteSizeLimitReached()) {
+                utils.debugLog(`Queue byte size reached ${this.currentByteSize} Bytes, releasing batch... `);
+                this.releaseBatch(this.batchSize - 1);
+            }
+        });
+
+        this.on('batchReleased', (batch) => {
+            this.batchSender(batch);
+            this.emit('batchSent', batch);
+        });
+    }
+
+    /**
+   * Queue size getter
+   * @returns {Number}
+   */
+    get currentSize() {
+        return this.traces.length;
+    }
+
+
+    /**
+   * Checks if queue size reached batch size
+   * @returns {Boolean}
+   */
+    batchSizeReached() {
+        return this.currentSize === this.batchSize;
+    }
+
+    /**
+   * Checks if queue byte size reached its limit
+   * @returns {Boolean}
+   */
+    byteSizeLimitReached() {
+        return this.currentByteSize >= this.batchBytesLimit;
+    }
+
+    /**
+   * add given trace byte size to total byte size
+   * @returns {Number}
+   */
+    addToCurrentByteSize(trace) {
+        this.currentByteSize += JSON.stringify(trace).length;
+    }
+
+    /**
+   * subtract given trace byte size to total byte size
+   * @returns {Number}
+   */
+    subtractFromCurrentByteSize(trace) {
+        this.currentByteSize -= JSON.stringify(trace).length;
+        this.currentByteSize = Math.max(this.currentByteSize, 0);
+    }
+
+    /**
+   * Flush queue
+   */
+    flush() {
+        this.traces = [];
+    }
+
+    /**
+   * Push trace to queue, emit event, and check if queue max queue length reached,
+   * if it does, send batch.
+   * @param {String} trace
+   * @returns {TraceQueue}
+   */
+    push(trace) {
+        const timestamp = new Date();
+        this.traces.push({ trace, timestamp });
+        utils.debugLog(`Trace ${trace} pushed to queue`);
+        this.emit('traceQueued', trace);
+        return this;
+    }
+
+
+    /**
+   * Release batch of traces
+   * @returns {TraceQueue}
+   */
+    releaseBatch(count = this.batchSize) {
```

Comment:
what's the added value of breaking `releaseBatch` & `batchReleased` to 2 different async actions? sounds like there is no reason to do it if we don't want concurrent sends.

Reply:
That is a good point, the issue here that we can have some kind of race condition... let's talk about that.